### PR TITLE
docker: fix a bug where auth for private registries wasn't parsed correctly

### LINKED ISSUE #24181: Docker auth.config not working anymore in 1.9.0
### Nomad version
```
Nomad v1.9.0
BuildDate 2024-10-10T07:13:43Z
Revision 7ad36851ec02f875e0814775ecf1df0229f0a615
```

### Operating system and Environment details
AlmaLinux 9, using pre-built amd64 bin

### Issue
Having docker plugin configured with an auth config like
```
plugin "docker" {
  config {
    auth {
      config = "/opt/nomad/docker/auth.json"
    }
[...]
```
With /opt/nomad/docker/auth.json looking like
```
{
  "auths": {
    "oci.ehtrace.local": {
      "auth": "XXXXXXX"
    }
  }
}
```

Doesn't work anymore : allocation fails to start with 
```
Failed to pull `oci.ehtrace.local/kafka-connect:3.8.0-45.0.0-SNAPSHOT-1`: Error response from daemon: Head "https://oci.ehtrace.local/v2/kafka-connect/manifests/3.8.0-45.0.0-SNAPSHOT-1": no basic auth credentials
```

The exact same setup was working fine in 1.8.4 (and previous versions)

### Reproduction steps
Try to start an allocation from a registrry requiring authentication, with credentials provided in auth.config

#### Expected Result
Credentials should be passed to Docker
### Job file (if appropriate)
Any job file is affected

### PR DESCRIPTION
**Pull request by pkazmierczak**
In https://github.com/hashicorp/nomad/pull/23966 we introduced an official Docker client and did not notice that in contrast to our previous 3rd party client, the official SDK `PullOptions` object expects a base64 encoded JSON with username and password, instead of username/password pair. 

This PR fixes the issue and adds e2e tests for Docker driver auth. 

Fixes #24181 
Internal ref: https://hashicorp.atlassian.net/browse/NET-11310

### CODE REVIEW

File: drivers/docker/utils.go [hunk 1]
```diff
@@ -102,7 +103,16 @@ func authFromTaskConfig(driverConfig *TaskConfig) authBackend {
 		if len(driverConfig.Auth.Username) == 0 && len(driverConfig.Auth.Password) == 0 && len(driverConfig.Auth.Email) == 0 && len(driverConfig.Auth.ServerAddr) == 0 {
 			return nil, nil
 		}
+		authConfig := registrytypes.AuthConfig{
+			Username: driverConfig.Auth.Username,
+			Password: driverConfig.Auth.Password,
+		}
+
+		encodedJSON, _ := json.Marshal(authConfig)
```
[the remark this answers is not in the document]
Comment (pkazmierczak) [reply drivers/docker/utils.go:111]: So the input we're getting here is already validated by helpers, and if there are malformed usernames/passwords they wouldn't get accepted by docker anyways, that's why I thought it's ok to 💀 here worst case scenario. But you're right, an additional error handling if statement won't hurt, I'll fix.

File: drivers/docker/utils.go [hunk 2]
```diff
@@ -140,6 +150,17 @@ func authFromDockerConfig(file string) authBackend {
...
+
+				// docker API calls require base64 encoded auth string
+				if auth.Username != "" && auth.Password != "" {
+					authConfig := registrytypes.AuthConfig{
+						Username: auth.Username,
+						Password: auth.Password,
+					}
+					encodedJSON, _ := json.Marshal(authConfig)
+					auth.Auth = base64.URLEncoding.EncodeToString(encodedJSON)
+				}
```
[the remark this answers is not in the document]
Comment (pkazmierczak) [reply drivers/docker/utils.go]: a1ae4c9

**Comment by pkazmierczak** [comment]
One other thing: the e2e tests are failing because the local registry isn't added to dockerd's "insecure registries." I'm not sure how to add it since we only know the registry address once we create it, I'll have to think about it.

File: drivers/docker/utils.go [hunk 3]
```diff
@@ -195,6 +208,21 @@ func authFromHelper(helperName string) authBackend {
 	}
 }
 
+// some docker api calls require a base64 encoded basic auth string
+func encodeAuth(cfg *registrytypes.AuthConfig, username, password string) error {
+	auth := &registrytypes.AuthConfig{
+		Username: username,
+		Password: password,
```
[the remark this answers is not in the document]
Comment (pkazmierczak) [reply drivers/docker/utils.go]: ooof, missed that 🤦 
[d0c89d8](https://github.com/hashicorp/nomad/pull/24215/commits/d0c89d8bda182031b0913e1941cf1eb987624154) fixes it

**Comment by shoenig** [comment]
The podman e2e test jumps through some fun hoops to setup a private registry that is auth enabled, can take a look for inspiration maybe. https://github.com/hashicorp/nomad/blob/main/e2e/podman/input/registry-auths.hcl

**Comment by gjrtimmer** [comment]
I had to rollback to `1.8.4`, this issues caused several hours of downtime. When is this going to be fixed?

**Review by tgross — APPROVED** [review]
LGTM, nice work on this

**Pull request merged**